source/MaterialXView/ShaderValidators/Glsl/GlslProgram.cpp

```diff
@@ -0,0 +1,644 @@
+
+#include <MaterialXView/External/GLew/glew.h>
+#include <MaterialXView/ShaderValidators/Glsl/GlslProgram.h>
+
+#include <iostream>
+#include <algorithm>
+#include <cmath>
+
+namespace MaterialX
+{
+// OpenGL Constants
+unsigned int GlslProgram::UNDEFINED_OPENGL_RESOURCE_ID = 0;
+int GlslProgram::UNDEFINED_OPENGL_PROGRAM_LOCATION = -1;
+int GlslProgram::Input::INVALID_OPENGL_TYPE = -1;
+
+//
+// Creator
+//
+GlslProgramPtr GlslProgram::creator()
+{
+    return std::shared_ptr<GlslProgram>(new GlslProgram());
+}
+
+GlslProgram::GlslProgram() :
+    _programId(UNDEFINED_OPENGL_RESOURCE_ID),
+    _hwShader(nullptr)
+{
+}
+
+GlslProgram::~GlslProgram()
+{
+    // Clean up the program and offscreen target
+    deleteProgram();
+}
+
+void GlslProgram::setStages(const std::vector<std::string>& stages)
+{
+    if (stages.size() != HwShader::NUM_STAGES)
+    {
+        ShaderValidationErrorList errors;
+        throw ExceptionShaderValidationError("Incorrect number of stages passed in for stage setup.", errors);
+    }
+
+    unsigned int count = 0;
+    for (auto stage : stages)
+    { 
+        _stages[count++] = stage;
+    }
+
+    // A stage change invalidates any cached parsed inputs
+    clearInputLists();
+    _hwShader = nullptr;
+}
+
+
+void GlslProgram::setStages(const HwShaderPtr shader)
+{
+    if (!shader)
+    {
+        ShaderValidationErrorList errors;
+        throw ExceptionShaderValidationError("Cannot set stages using null hardware shader.", errors);
+    }
+
+    // Clear out any old data
+    clearStages();
+
+    // Extract out the shader code per stage
+    _hwShader = shader;
+    for (size_t i = 0; i < HwShader::NUM_STAGES; i++)
+    {
+        _stages[i] = _hwShader->getSourceCode(i);
+    }
+
+    // A stage change invalidates any cached parsed inputs
+    clearInputLists();
+}
+
+const std::string GlslProgram::getStage(size_t stage) const
+{
+    if (stage < HwShader::NUM_STAGES)
+    {
+        return _stages[stage];
+    }
+    return std::string();
```

Comment:
Return EMPTY_STRING instead. Then this method can return a reference to string instead of a copy, and avoid copy the whole stage source.